### PR TITLE
Support adding properties during build promotion

### PR DESCRIPTION
**Pull request by AlexeiVainshtein**

### CODE REVIEW

File: jfrog-cli/artifactory/cli.go [hunk 1]
```diff
@@ -909,14 +909,18 @@ func getSearchFlags() []cli.Flag {
...
-		cli.StringFlag{
-			Name:  "props",
-			Usage: "[Optional] List of properties in the form of \"key1=value1;key2=value2,...\". Only artifacts with these properties are affected.` `",
-		},
+		getPropertiesFlag("Only artifacts with these properties are affected."),
 	}
 	return append(flags, getPropertiesFlags()...)
 }
 
+func getPropertiesFlag(description string) cli.Flag {
```
Comment (eyalbe4) [inline jfrog-cli/artifactory/cli.go]: It looks like this new function is not used to create the props option for **all** commands. See move, copy, delete and more...

File: jfrog-cli/jfrog/buildinfo_test.go [hunk 2]
```diff
@@ -38,6 +41,85 @@ func TestBuildAddDependenciesFromHomeDir(t *testing.T) {
...
+
+	if len(buildInfo.Modules[0].Artifacts) != len(resultItems) {
+		t.Error("Incorrect number of artifacts were uploaded, expected:", len(buildInfo.Modules[0].Artifacts), " Found:", len(resultItems))
+	}
+
+	propsMap := make(map[string]string)
+	propsMap["build.name"] = buildInfo.Name
+	propsMap["build.number"] = buildInfo.Number
+	propsMap[key1] = value1
+	propsMap[key2] = value2
```
Comment (eyalbe4) [inline jfrog-cli/jfrog/buildinfo_test.go]: A map can be also initialised as follows (shorter syntax):
```
propsMap := map[string]string{
    "rsc": 3711,
    "r":   2138,
    "gri": 1908,
    "adg": 912,
}
```

File: jfrog-cli/jfrog/buildinfo_test.go [hunk 3]
```diff
@@ -38,6 +41,85 @@ func TestBuildAddDependenciesFromHomeDir(t *testing.T) {
...
+
+	validateArtifactsProperties(resultItems, t, propsMap)
+	cleanArtifactoryTest()
+}
+
+// Returns the artifacts result found by the provided spec
```
Comment (eyalbe4) [inline jfrog-cli/jfrog/buildinfo_test.go]: // Returns the artifacts **result** found by the provided spec --> // Returns the artifacts found by the provided spec

File: jfrog-cli/jfrog/buildinfo_test.go [hunk 4]
```diff
@@ -38,6 +41,85 @@ func TestBuildAddDependenciesFromHomeDir(t *testing.T) {
...
+		currentResultItems, err := services.SearchBySpecFiles(searchParams, flags, rtutils.ALL)
+		if err != nil {
+			t.Error("Failed Searching files:", err)
+		}
+		resultItems = append(resultItems, currentResultItems...)
+	}
+	return resultItems
+}
+
+// This function validates the properties on the provided artifacts. Every property within the provided map should appear on the artifact.:
```
Comment (eyalbe4) [inline jfrog-cli/jfrog/buildinfo_test.go]: ...should appear on the artifact.: --> ...should be attached to the artifact.

File: jfrog-cli/jfrog/buildinfo_test.go [hunk 5]
```diff
@@ -38,6 +41,85 @@ func TestBuildAddDependenciesFromHomeDir(t *testing.T) {
...
+func validateArtifactsProperties(resultItems []rtutils.ResultItem, t *testing.T, propsMap map[string]string) {
+	for _, item := range resultItems {
+		properties := item.Properties
+		if len(properties) < 1 {
+			t.Error("Failed finding properties on item:", item.GetItemRelativePath())
+		}
+		propertiesMap := convertSliceToMap(properties)
```
Comment (eyalbe4) [inline jfrog-cli/jfrog/buildinfo_test.go]: The *convertSliceToMap* function is currently part of the *buildtools_test.go*, but is now also used by this file. I therefore suggest moving it to *utils/tests/utils.go*.

File: jfrog-cli/jfrog/buildtools_test.go [hunk 6]
```diff
@@ -270,7 +270,16 @@ func TestGoBuildInfo(t *testing.T) {
...
-	validateBuildInfoProperties(buildInfo, t)
+	resultItems := getResultItemsFromArtifactory(tests.SearchGo, t)
+	if len(buildInfo.Modules[0].Artifacts) != len(resultItems) {
+		t.Error("Incorrect number of artifacts were uploaded, expected:", len(buildInfo.Modules[0].Artifacts), " Found:", len(resultItems))
+	}
+	propsMap := make(map[string]string)
+	propsMap["build.name"] = buildName
+	propsMap["build.number"] = buildNumber
+	propsMap["go.version"] = "v1.0.0"
+	propsMap["go.name"] = "github.com/jfrog/dependency"
```
Comment (eyalbe4) [inline jfrog-cli/jfrog/buildtools_test.go]: A map can be also initialised as follows (shorter syntax):
```
propsMap := map[string]string{
    "rsc": 3711,
    "r":   2138,
    "gri": 1908,
    "adg": 912,
}
```

File: jfrog-cli/artifactory/cli.go [hunk 7]
```diff
@@ -1028,7 +1032,7 @@ func getBuildPromotionFlags() []cli.Flag {
 		cli.BoolFlag{
 			Name:  "dry-run",
 			Usage: "[Default: false] If true, promotion is only simulated. The build is not promoted.` `",
-		},
+		}, getPropertiesFlag("A list of properties to attach to the build's artifacts."),
```
[the remark this answers is not in the document]
Comment (AlexeiVainshtein) [reply jfrog-cli/artifactory/cli.go]: Done

File: jfrog-cli/jfrog/buildinfo_test.go (hunk 2 above)
Comment (AlexeiVainshtein) [reply jfrog-cli/jfrog/buildinfo_test.go]: Done

File: jfrog-cli/jfrog/buildinfo_test.go (hunk 3 above)
Comment (AlexeiVainshtein) [reply jfrog-cli/jfrog/buildinfo_test.go]: Done

File: jfrog-cli/jfrog/buildinfo_test.go (hunk 4 above)
Comment (AlexeiVainshtein) [reply jfrog-cli/jfrog/buildinfo_test.go]: Done

File: jfrog-cli/jfrog/buildinfo_test.go (hunk 5 above)
Comment (AlexeiVainshtein) [reply jfrog-cli/jfrog/buildinfo_test.go]: Done

File: jfrog-cli/jfrog/buildtools_test.go (hunk 6 above)
Comment (AlexeiVainshtein) [reply jfrog-cli/jfrog/buildtools_test.go]: Done

File: jfrog-cli/artifactory/cli.go (hunk 1 above)
Comment (AlexeiVainshtein) [reply jfrog-cli/artifactory/cli.go]: Done

**Pull request merged**